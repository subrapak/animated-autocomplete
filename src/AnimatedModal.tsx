import React, {
  Dispatch,
  FunctionComponent,
  RefObject,
  useEffect,
  useState,
} from 'react';
import {
  Dimensions,
  FlatList,
  Image,
  Keyboard,
  StyleSheet,
  Text,
  TextInput,
  TouchableOpacity,
  View,
} from 'react-native';
import Animated, { Easing, Value } from 'react-native-reanimated';

const { width, height } = Dimensions.get('window');

type Props = {
  translateY: Value<number>;
  animatedHeight: Value<number>;
  datalist: string[];
  autocompleteText: string;
  setAutocompleteText: Dispatch<React.SetStateAction<string>>;
  closeIcon?: JSX.Element;
  searchIcon?: JSX.Element;
  searchRef: RefObject<TextInput>;
};

export const AnimatedModal: FunctionComponent<Props> = ({
  translateY,
  animatedHeight,
  datalist,
  autocompleteText,
  setAutocompleteText,
  closeIcon,
  searchIcon,
  searchRef,
}: Props) => {
  const listMaxHeight = new Value<number>(height * 0.8);

  const closeModal = (item?: string) => {
    Animated.timing(translateY, {
      toValue: height,
      duration: 300,
      easing: Easing.ease,
    }).start(() => {
      item && setAutocompleteText(item);
      setSearchBarText('');
      setFilterData(originalData);
      Keyboard.dismiss();
    });
    Animated.timing(animatedHeight, {
      toValue: 0,
      duration: 500,
      easing: Easing.ease,
    }).start();
  };
  useEffect(() => {
    Keyboard.addListener('keyboardDidShow', _keyboardDidShow);
    Keyboard.addListener('keyboardDidHide', _keyboardDidHide);

    // cleanup function
    return () => {
      Keyboard.removeListener('keyboardDidShow', _keyboardDidShow);
      Keyboard.removeListener('keyboardDidHide', _keyboardDidHide);
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  const _keyboardDidShow = () => {
    Animated.timing(listMaxHeight, {
      toValue: height * 0.5,
      duration: 500,
      easing: Easing.ease,
    }).start();
  };

  const _keyboardDidHide = () => {
    Animated.timing(listMaxHeight, {
      toValue: height * 0.8,
      duration: 300,
      easing: Easing.ease,
    }).start();
  };

  const renderItem = ({ item }: { item: string }) => {
    return (
      <TouchableOpacity
        style={styles.touchableOpacity}
        onPress={() => closeModal(item)}
      >
        <Text style={styles.renderText}>{item}</Text>
      </TouchableOpacity>
    );
  };

  const [originalData] = useState<string[]>(datalist);
  const [filterData, setFilterData] = useState<string[]>(datalist);
  const [searchBarText, setSearchBarText] = useState<string>(autocompleteText);
  const handleChangeSearchBarText = (text: string) => {
    setSearchBarText(text);

    text.length
      ? setFilterData(originalData.filter((item) => item.includes(text)))
      : setFilterData(originalData);
  };

  return (
    <Animated.View
      style={{
        ...styles.animatedModal,
        height: animatedHeight,
        transform: [{ translateY: translateY }],
      }}
    >
      <View style={styles.mainContent}>
        <View style={styles.searchBar}>
          {searchIcon ?? (
            <Image
              source={require('./assets/search-outline.jpg')}
              style={styles.searchIcon}
            />
          )}
          <TextInput
            style={styles.textInput}
            value={searchBarText}
            onChangeText={handleChangeSearchBarText}
            autoCorrect={false}
            ref={searchRef}
          />
          <TouchableOpacity onPress={() => closeModal()}>
            {closeIcon ?? (
              <Image
                source={require('./assets/close-outline.jpg')}
                style={styles.closeIcon}
              />
            )}
          </TouchableOpacity>
        </View>
        <Animated.View style={[styles.animatedView, { height: listMaxHeight }]}>
          <FlatList
            data={filterData}
            renderItem={renderItem}
            keyExtractor={(item: string) => item}
            keyboardShouldPersistTaps={'always'}
          />
        </Animated.View>
      </View>
    </Animated.View>
  );
};

const styles = StyleSheet.create({
  searchBar: {
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'space-around',
    borderRadius: 10,
    shadowOffset: { width: 2, height: 2 },
    shadowColor: 'black',
    shadowRadius: 5,
    shadowOpacity: 0.5,
    backgroundColor: 'white',
    padding: 10,
  },
  searchIcon: {
    height: 30,
    width: 30,
  },
  closeIcon: {
    height: 30,
    width: 30,
  },
  renderText: { fontSize: 18, color: 'grey' },
  animatedView: {
    overflow: 'hidden',
    paddingVertical: 5,
  },
  iconStyle: {
    display: 'flex',
    justifyContent: 'center',
    alignItems: 'center',
    padding: 0,
  },
  mainContent: {
    width: '100%',
    height: '100%',
    padding: 10,
  },
  textInput: {
    width: '80%',
    height: '100%',
    backgroundColor: 'white',
    borderColor: 'white',
    borderWidth: 1,
    fontSize: 25,
    padding: 0,
  },
  touchableOpacity: {
    width: '100%',
    backgroundColor: '#f2f2f2',
    borderBottomColor: 'grey',
    borderBottomWidth: 1,
    padding: 15,
  },
  animatedModal: {
    position: 'absolute',
    top: 0,
    height: height,
    width: width,
    marginHorizontal: 'auto',
    display: 'flex',
    borderRadius: 25,
    paddingVertical: 50,
    justifyContent: 'flex-start',
    backgroundColor: 'white',
    alignItems: 'center',
  },
});
