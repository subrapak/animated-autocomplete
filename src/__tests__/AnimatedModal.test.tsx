import { AnimatedModal } from '../AnimatedModal';
import { render } from '@testing-library/react-native';
import { Value } from 'react-native-reanimated';
import React, { createRef, useState } from 'react';
import type { TextInput } from 'react-native';

const mockModalHeight = new Value<number>(0);
const mockModalTranslateY = new Value<number>(848);
const mockData = ['mock1', 'mock2'];
const mockSearchBarRef = createRef<TextInput>();
jest.mock('react-native-reanimated', () =>
  require('react-native-reanimated/mock')
);
describe('Testing the Animated Modal component', () => {
  it('should render the animated modal component properly', () => {
    const animatedModal = render(
      <AnimatedModal
        animatedHeight={mockModalHeight}
        translateY={mockModalTranslateY}
        datalist={mockData}
        autocompleteText={'mockMainText'}
        setAutocompleteText={() => jest.fn()}
        searchRef={mockSearchBarRef}
      />
    );
    expect(animatedModal).toMatchSnapshot();
  });
});
