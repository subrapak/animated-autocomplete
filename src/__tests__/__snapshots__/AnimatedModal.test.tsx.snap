// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the Animated Modal component should render the animated modal component properly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "white",
      "borderRadius": 25,
      "display": "flex",
      "height": AnimatedValue {
        " __value": 0,
      },
      "justifyContent": "flex-start",
      "marginHorizontal": "auto",
      "paddingVertical": 50,
      "position": "absolute",
      "top": 0,
      "transform": Array [
        Object {
          "translateY": AnimatedValue {
            " __value": 848,
          },
        },
      ],
      "width": 750,
    }
  }
>
  <View
    style={
      Object {
        "height": "100%",
        "padding": 10,
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderRadius": 10,
          "flexDirection": "row",
          "justifyContent": "space-around",
          "padding": 10,
          "shadowColor": "black",
          "shadowOffset": Object {
            "height": 2,
            "width": 2,
          },
          "shadowOpacity": 0.5,
          "shadowRadius": 5,
        }
      }
    >
      <Image
        source={
          Object {
            "testUri": "../../../src/assets/search-outline.jpg",
          }
        }
        style={
          Object {
            "height": 30,
            "width": 30,
          }
        }
      />
      <TextInput
        allowFontScaling={true}
        autoCorrect={false}
        onChangeText={[Function]}
        rejectResponderTermination={true}
        style={
          Object {
            "backgroundColor": "white",
            "borderColor": "white",
            "borderWidth": 1,
            "fontSize": 25,
            "height": "100%",
            "padding": 0,
            "width": "80%",
          }
        }
        underlineColorAndroid="transparent"
        value="mockMainText"
      />
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <Image
          source={
            Object {
              "testUri": "../../../src/assets/close-outline.jpg",
            }
          }
          style={
            Object {
              "height": 30,
              "width": 30,
            }
          }
        />
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "overflow": "hidden",
            "paddingVertical": 5,
          },
          Object {
            "height": AnimatedValue {
              " __value": 1067.2,
            },
          },
        ]
      }
    >
      <RCTScrollView
        data={
          Array [
            "mock1",
            "mock2",
          ]
        }
        disableVirtualization={false}
        getItem={[Function]}
        getItemCount={[Function]}
        horizontal={false}
        initialNumToRender={10}
        keyExtractor={[Function]}
        keyboardShouldPersistTaps="always"
        maxToRenderPerBatch={10}
        onContentSizeChange={[Function]}
        onEndReachedThreshold={2}
        onLayout={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={50}
        stickyHeaderIndices={Array []}
        updateCellsBatchingPeriod={50}
        viewabilityConfigCallbackPairs={Array []}
        windowSize={21}
      >
        <View>
          <View
            onLayout={[Function]}
            style={null}
          >
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "backgroundColor": "#f2f2f2",
                  "borderBottomColor": "grey",
                  "borderBottomWidth": 1,
                  "opacity": 1,
                  "padding": 15,
                  "width": "100%",
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "grey",
                    "fontSize": 18,
                  }
                }
              >
                mock1
              </Text>
            </View>
          </View>
          <View
            onLayout={[Function]}
            style={null}
          >
            <View
              accessible={true}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "backgroundColor": "#f2f2f2",
                  "borderBottomColor": "grey",
                  "borderBottomWidth": 1,
                  "opacity": 1,
                  "padding": 15,
                  "width": "100%",
                }
              }
            >
              <Text
                style={
                  Object {
                    "color": "grey",
                    "fontSize": 18,
                  }
                }
              >
                mock2
              </Text>
            </View>
          </View>
        </View>
      </RCTScrollView>
    </View>
  </View>
</View>
`;
